---
title: "Export data from seurat object slots"
output: html_notebook
---



# Export 10X for Seaborn clustering
## Load libraries required for Seuratv4
```{r message=FALSE, warning=FALSE}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/")
knitr::opts_chunk$set(fig.width = 6, fig.height = 4)
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(clustree)
```
## Source functions
```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/Mouse2Human_idconversion.R")
```

## Set global variables
```{r}
pop.id <- "CMP"
folder.appendix <- "m2"
fav.var <- 90
object.res <- ".0.5"
ident.list <- c("RNA_snn_res.0.5", "RNA_snn_res.1")
folder.id <- paste0(pop.id, folder.appendix)
projectName <- paste0(pop.id, "Subpop_vBiomark")
```

## Store session info
```{r}
sessionInfo.filename <- paste0(projectName, "_sessionInfo.txt")
sink(sessionInfo.filename)
sessionInfo()
sink()
```

*STOP* - testing showed that cells cluster better in seaborn when you include all probe data points. It was an intersting idea to cluster based only on the genes in the biomark probe set, but this doesn't seem to hold up in practice

```{r warning=FALSE, message=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = c(folder.id))
data.object<-Read10X(data_file.list)
seurat.object<- CreateSeuratObject(counts = data.object, min.cells = 3, min.genes = 200, project = projectName)
remove(data.object)
probe.list <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/BiomarkRNAs.txt", header = FALSE)
probe.list <- unlist(as.list(probe.list), use.names = FALSE)
probe.list <- c(probe.list, "Kit", "Il3ra")
# Subset to biomark features
biomark.object <- subset(seurat.object, features = probe.list)
biomark.object[["percent.mt"]] <- PercentageFeatureSet(biomark.object, pattern = "^mt-")
print(paste("original object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- subset(biomark.object, 
												 subset = nFeature_RNA >=25 & 
												 	percent.mt <= 5
)
print(paste("new object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- NormalizeData(biomark.object, normalization.method = "LogNormalize", scale.factor = 10000)
object.downsample <- subset(biomark.object, cells = sample(Cells(biomark.object), 300))
```



## overlay data from clustered seurat.object onto biomark.object
```{r}
seurat.object <- readRDS("~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/CMPSubpop_pctVar90.RDS")
xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "counts"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Counts-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Data-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "scale.data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_ScaleData-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

```








