---
title: "Mergebiomark167With10X"
output:
  html_document:
    df_print: paged
---

# Updates

## 20211005
Compared to archive 1, this version imports Biomark Ct spreadsheets that do not include fluorophore data 

```{r setup}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/MergeBiomarkWith10X/")
```

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(clustree)
```

## Set global variables
```{r}
fav.var <- 95
object.res <- ".0.5"
ident.list <- c("integrated_snn_res.0.5", "integrated_snn_res.1")
projectName <- "Mergebiomark167With10X"
```

```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/Mouse2Human_idconversion.R")
```
store session info
```{r }
sink(paste0(projectName, ".txt"))
sessionInfo()
sink()
```

# Create seurat object

## Load biomark data
Try to read in Biomark Data

```{r}
biomark.data <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/Biomark_CT.txt", sep = "\t", header = TRUE, row.names = 1)
biomark.data[is.na(biomark.data)] <- 0

# Limit to CMP data
biomark.data <- biomark.data[grepl("CMP", rownames(biomark.data)) == TRUE,]
dim(biomark.data)
```
```{r}
# fix non-MGI gene names
colnames(biomark.data)[colnames(biomark.data) == 'Cd117'] <- 'Kit'
colnames(biomark.data)[colnames(biomark.data) == 'Cd123'] <- 'Il3ra'
colnames(biomark.data)[colnames(biomark.data) == 'H2.Aa'] <- 'H2-Aa'
colnames(biomark.data)[colnames(biomark.data) == 'H2.Ab1'] <- 'H2-Ab1'
biomark.data <- subset(biomark.data, select = -Kng2)
```
Convert to pseudoexpression data
```{r}
# generate list of reference Ct averages, calculating average for ref genes where ref_gene Ct > 0
ref_list <- c("Actb", "B2m", "Kit")
biomark.data["refCt"] <- NA

for(i in 1:nrow(biomark.data)){
	ref_cts <- c()
	for(j in ref_list){
		if(biomark.data[i, j] > 0){
			ref_cts <- c(ref_cts, biomark.data[i, j])
		}
	}
	biomark.data[i, "refCt"] <- mean(ref_cts)
}

# generate dCt col
dbiomark.data <- biomark.data[, 1:ncol(biomark.data) -1]
dbiomark.data[dbiomark.data == 0] <- NA
dbiomark.data <- dbiomark.data - biomark.data[, "refCt"]

# create pseudo expression matrix
pseudo.biomark <- 2^-dbiomark.data

```





```{r warning=FALSE, message=FALSE}
biomark.meta <- data.frame(stringr::str_split_fixed(rownames(pseudo.biomark), "_", 3)[,2:3], row.names = rownames(pseudo.biomark))
colnames(biomark.meta) <- c("wid", "pid")

pseudo.biomark.T <- data.table::transpose(pseudo.biomark)
colnames(pseudo.biomark.T) <- rownames(pseudo.biomark)
rownames(pseudo.biomark.T) <- colnames(pseudo.biomark)

biomark.meta['nFeature_RNA'] <- NA
for(cell in colnames(pseudo.biomark.T)){
	ngenes <- length(which(!(pseudo.biomark.T[colnames(pseudo.biomark.T > 0) == cell])))
	biomark.meta[rownames(biomark.meta) == cell, 'nFeature_RNA'] <- ngenes
}

```

```{r}
pseudo.biomark.T[is.na(pseudo.biomark.T)] <- 0
pseudo.biomark.T <- as.sparse(pseudo.biomark.T)
biomark <- CreateSeuratObject(pseudo.biomark.T, project = "Biomark", assay = "RNA", meta.data = biomark.meta)
```

## Load SC data

```{r warning=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = "CMPm2")
object.data <-Read10X(data_file.list)
cmp.object<- CreateSeuratObject(counts = object.data, min.cells = 3, min.genes = 200, project = "10X")
```


Remove non-biomark genes from cmp.object

```{r}
biomark.genes <- rownames(biomark@assays$RNA@data)
cmp.object <- subset(cmp.object, features = biomark.genes)
cmp.object@meta.data["wid"] <- '10X'
cmp.object@meta.data["pid"] <- "p14"
setdiff(rownames(biomark@assays$RNA@data), rownames(cmp.object@assays$RNA@data))
```

Clean up to free memory

```{r}
remove(object.data, biomark.data, biomark.meta)
```

## merge objects
```{r}
seurat.object <- merge(cmp.object, biomark, project = 'mergebiomark')
seurat.object@meta.data$orig.ident[seurat.object@meta.data$orig.ident == 'CMPm2'] <- '10X'
seurat.object@meta.data$orig.ident[seurat.object@meta.data$orig.ident == 'CMP'] <- 'Biomark'
# seurat.object@meta.data[is.na(seurat.object@meta.data)] <- 0
# seurat.object@meta.data$orig.ident <- factor(seurat.object@meta.data$orig.ident)
# seurat.object@meta.data$pid <- factor(seurat.object@meta.data$pid)
# seurat.object@meta.data$wid <- factor(seurat.object@meta.data$wid)
seurat.object
```

# Traditional normalize and scaling approaches
trying sctransform and normalizeData/scaledata
```{r}
# seurat.object <- SCTransform(seurat.object, vars.to.regress = c("wid"))
seurat.object <- NormalizeData(seurat.object, normalization.method = "LogNormalize", scale.factor = 10000)
seurat.object <- FindVariableFeatures(seurat.object, selection.method = "vst", nfeatures = 10)
```

Scale data (linear transformation)

```{r}
all.genes <- rownames(seurat.object)
seurat.object <- ScaleData(seurat.object, features = all.genes, vars.to.regress = c("wid", "pid", "orig.ident"))
```

### Save merged object
```{r}
saveRDS(seurat.object, file = paste0(projectName, "_raw.RDS"))
```
## Run PCA
```{r}
seurat.object <- RunPCA(seurat.object, ndims.print = 1:5, nfeatures.print = 5, features = all.genes)
```


```{r}
DimPlot(seurat.object, reduction = "pca", group.by = "orig.ident")
```

```{r}
VizDimLoadings(seurat.object, dims = 1:6, nfeatures = 10, reduction = "pca", ncol = 3)

```

Calculate dimensionality
```{r, figures-side, fig.show='hold', out.width="50%"}
ElbowPlot(seurat.object, ndims = 50)
percent.variance(seurat.object@reductions$pca@stdev)
```




## Calculate percent variance
```{r}
tot.var <- percent.variance(seurat.object@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance:", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance:", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance:", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance:", length(which(cumsum(tot.var) <= 95)))

```

# scRNASeq integration	
neither sctransform nor normalizedata/scaledata gave what looked like reasonable results.

https://satijalab.org/seurat/articles/integration_introduction.html


```{r}
tech.list <- c(cmp.object, biomark)
tech.list <- lapply(X = tech.list, FUN = function(x) {
	x <- NormalizeData(x)
	x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 10)
})
```


## identify anchors
```{r}

tech.anchors <- FindIntegrationAnchors(object.list = tech.list, anchor.features = all.genes, k.anchor = 20, k.filter = NA)
# tech.integrated <- IntegrateData(anchorset = tech.anchors, dims = 1:30, k.weight = 20, normalization.method = "SCT")
tech.integrated <- IntegrateData(anchorset = tech.anchors, k.weight = 5)
```
## tech.integrated PCA

```{r}
DefaultAssay(tech.integrated) <- "integrated"
tech.integrated <- ScaleData(tech.integrated)
tech.integrated <- RunPCA(tech.integrated, npcs = 30)
DimPlot(tech.integrated, reduction = "pca", group.by = "orig.ident", pt.size = 2)
```

so it looks like anchor-based integration is much nicer! goign to stick with this approach.



Number of PCs describing X% of variance

```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance: ", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance: ", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance: ", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance: ", length(which(cumsum(tot.var) <= 95)))

```


# ID clusters based on different variances
## Describe 80% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 80
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```







## Describe 85% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 85
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```






## Describe 90% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 90
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```






## Describe 95% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 95
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r warning=FALSE, results='hide'}
for(x in c(0.5, 1, 1.5, 2, 2.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, " pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```



## Create `FindAllMarkers()` lists for GSEA
```{r}
ident.list <- c("integrated_snn_res.0.5")

for(tested.ident in ident.list){
	Idents(tech.integrated) <- tested.ident
	all.markers <- FindAllMarkers(tech.integrated)
	xlsx::write.xlsx(x = all.markers[,c("avg_log2FC", "p_val_adj", "cluster", "gene")], 
									 file = paste0(projectName, " pctVar", pct.var, "_allres.xlsx"), 
									 sheetName = tested.ident, 
									 col.names = TRUE, 
									 row.names = FALSE, 
									 append = TRUE)
}
```

## Map HGNC symbols
```{r}
Mouse2HumanTable <- Mouse2Human(object.res.allmarkers$gene)

HGNC <- with(Mouse2HumanTable, Mouse2HumanTable$HGNC[match(object.res.allmarkers$gene, Mouse2HumanTable$MGI)])
head(object.res.allmarkers)
object.res.allmarkers$HGNC <- HGNC
tail(object.res.allmarkers)
sig.res <- object.res.allmarkers[object.res.allmarkers$p_val_adj <= 0.05, ]
sig.res <- sig.res[c("avg_log2FC", "HGNC", "cluster")]
sig.res <- sig.res[!(sig.res$HGNC == "" | is.na(sig.res$HGNC)),] # GSEA will fail if there are any blanks or NAs in the table
sig.res <- sig.res[]

```


```{r}
for(cluster in unique(sig.res$cluster)){
	print(paste("writing cluster", cluster))
	new.table <- sig.res[sig.res$cluster == cluster, c("HGNC", "avg_log2FC")]
	new.table <- new.table[order(-new.table$avg_log2FC), ]
	try(dir.create(paste0("RankList_res", object.res, "_findAll_hgnc/"), showWarnings = FALSE))
	try(write.table(new.table, file = paste0(projectName, "RankList_pctVar", pct.var, "res", object.res, "_findAll_hgnc/res", object.res, "cluster", cluster, ".rnk"), quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t", ))
	
}
```




# Export 10X for Seaborn clustering
## Load libraries required for Seuratv4
```{r message=FALSE, warning=FALSE}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/")
knitr::opts_chunk$set(fig.width = 6, fig.height = 4)
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(clustree)
```
## Source functions
```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/Mouse2Human_idconversion.R")
```

## Set global variables
```{r}
pop.id <- "CMP"
folder.appendix <- "m2"
fav.var <- 90
object.res <- ".0.5"
ident.list <- c("RNA_snn_res.0.5", "RNA_snn_res.1")
folder.id <- paste0(pop.id, folder.appendix)
projectName <- paste0(pop.id, "Subpop_vBiomark")
```

## Store session info
```{r}
sessionInfo.filename <- paste0(projectName, "_sessionInfo.txt")
sink(sessionInfo.filename)
sessionInfo()
sink()
```

*STOP* - testing showed that cells cluster better in seaborn when you include all probe data points. It was an intersting idea to cluster based only on the genes in the biomark probe set, but this doesn't seem to hold up in practice

```{r warning=FALSE, message=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = c(folder.id))
data.object<-Read10X(data_file.list)
seurat.object<- CreateSeuratObject(counts = data.object, min.cells = 3, min.genes = 200, project = projectName)
remove(data.object)
probe.list <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/BiomarkRNAs.txt", header = FALSE)
probe.list <- unlist(as.list(probe.list), use.names = FALSE)
probe.list <- c(probe.list, "Kit", "Il3ra")
# Subset to biomark features
biomark.object <- subset(seurat.object, features = probe.list)
biomark.object[["percent.mt"]] <- PercentageFeatureSet(biomark.object, pattern = "^mt-")
print(paste("original object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- subset(biomark.object, 
												 subset = nFeature_RNA >=25 & 
												 	percent.mt <= 5
)
print(paste("new object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- NormalizeData(biomark.object, normalization.method = "LogNormalize", scale.factor = 10000)
object.downsample <- subset(biomark.object, cells = sample(Cells(biomark.object), 300))
```



## overlay data from clustered seurat.object onto biomark.object
```{r}
seurat.object <- readRDS("~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/CMPSubpop_pctVar90.RDS")
xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "counts"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Counts-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Data-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "scale.data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_ScaleData-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

```




```{r}
biomark.object <- readRDS("~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/CMPSubpop_pctVar90.RDS")
```








