---
title: "Mergebiomark167With10X"
output:
  html_document:
    df_print: paged
---

# Updates

## 20211005
Compared to archive 1, this version imports Biomark Ct spreadsheets that do not include fluorophore data 

```{r setup}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/MergeBiomarkWith10X/")
```

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
```

# Set global variables
```{r}
projectName <- "Mergebiomark167With10X"
```



```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
```


store session info
```{r }
sink(paste0(projectName, ".txt"))
sessionInfo()
sink()
```

# Create seurat object

## Load biomark data
Try to read in Biomark Data

```{r}
biomark.data <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/Biomark_CT.txt", sep = "\t", header = TRUE, row.names = 1)
biomark.data[is.na(biomark.data)] <- 0

# Limit to CMP data
biomark.data <- biomark.data[grepl("CMP", rownames(biomark.data)) == TRUE,]
dim(biomark.data)
```
Convert to pseudoexpression data
```{r}
# generate list of reference Ct averages, calculating average for ref genes where ref_gene Ct > 0
ref_list <- c("Actb", "B2m", "Kit")
biomark.data["refCt"] <- NA

for(i in 1:nrow(biomark.data)){
	ref_cts <- c()
	for(j in ref_list){
		if(biomark.data[i, j] > 0){
			ref_cts <- c(ref_cts, biomark.data[i, j])
		}
	}
	biomark.data[i, "refCt"] <- mean(ref_cts)
}

# generate dCt col
dbiomark.data <- biomark.data[, 1:ncol(biomark.data) -1]
dbiomark.data[dbiomark.data == 0] <- NA
dbiomark.data <- dbiomark.data - biomark.data[, "refCt"]

# create pseudo expression matrix
pseudo.biomark <- 2^-dbiomark.data

```


```{r}
# fix non-MGI gene names
colnames(pseudo.biomark)[colnames(pseudo.biomark) == 'Cd117'] <- 'Kit'
colnames(pseudo.biomark)[colnames(pseudo.biomark) == 'Cd123'] <- 'Il3ra'
colnames(pseudo.biomark)[colnames(pseudo.biomark) == 'H2.Aa'] <- 'H2-Aa'
colnames(pseudo.biomark)[colnames(pseudo.biomark) == 'H2.Ab1'] <- 'H2-Ab1'
pseudo.biomark <- subset(pseudo.biomark, select = -Kng2)
```



```{r warning=FALSE, message=FALSE}
biomark.meta <- data.frame(stringr::str_split_fixed(rownames(pseudo.biomark), "_", 3)[,2:3], row.names = rownames(pseudo.biomark))
colnames(biomark.meta) <- c("wid", "pid")

pseudo.biomark.T <- data.table::transpose(pseudo.biomark)
colnames(pseudo.biomark.T) <- rownames(pseudo.biomark)
rownames(pseudo.biomark.T) <- colnames(pseudo.biomark)

biomark.meta['nFeature_RNA'] <- NA
for(cell in colnames(pseudo.biomark.T)){
	ngenes <- length(which(!(pseudo.biomark.T[colnames(pseudo.biomark.T > 0) == cell])))
	biomark.meta[rownames(biomark.meta) == cell, 'nFeature_RNA'] <- ngenes
}

```

```{r}
pseudo.biomark.T[is.na(pseudo.biomark.T)] <- 0
pseudo.biomark.T <- as.sparse(pseudo.biomark.T)
biomark <- CreateSeuratObject(pseudo.biomark.T, project = "Biomark", assay = "RNA", meta.data = biomark.meta)
```

## Load SC data

```{r warning=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = "CMPm2")
object.data <-Read10X(data_file.list)
cmp.object<- CreateSeuratObject(counts = object.data, min.cells = 3, min.genes = 200, project = "10X")
```


Remove non-biomark genes from cmp.object

```{r}
biomark.genes <- rownames(biomark@assays$RNA@data)
cmp.object <- subset(cmp.object, features = biomark.genes)
cmp.object@meta.data["wid"] <- '10X'
cmp.object@meta.data["pid"] <- "p14"
setdiff(rownames(biomark@assays$RNA@data), rownames(cmp.object@assays$RNA@data))
```

Clean up to free memory

```{r}
remove(object.data, biomark.data, biomark.meta)
```

## merge objects
```{r}
seurat.object <- merge(cmp.object, biomark, project = 'mergebiomark')
seurat.object@meta.data$orig.ident[seurat.object@meta.data$orig.ident == 'CMPm2'] <- '10X'
seurat.object@meta.data$orig.ident[seurat.object@meta.data$orig.ident == 'CMP'] <- 'Biomark'
# seurat.object@meta.data[is.na(seurat.object@meta.data)] <- 0
# seurat.object@meta.data$orig.ident <- factor(seurat.object@meta.data$orig.ident)
# seurat.object@meta.data$pid <- factor(seurat.object@meta.data$pid)
# seurat.object@meta.data$wid <- factor(seurat.object@meta.data$wid)
seurat.object
```

# Traditional normalize and scaling approaches
trying sctransform and normalizeData/scaledata
```{r}
# seurat.object <- SCTransform(seurat.object, vars.to.regress = c("wid"))
seurat.object <- NormalizeData(seurat.object, normalization.method = "LogNormalize", scale.factor = 10000)
seurat.object <- FindVariableFeatures(seurat.object, selection.method = "vst", nfeatures = 10)
```

Scale data (linear transformation)

```{r}
all.genes <- rownames(seurat.object)
seurat.object <- ScaleData(seurat.object, features = all.genes, vars.to.regress = c("wid", "pid", "orig.ident"))
```

### Save merged object
```{r}
saveRDS(seurat.object, file = paste0(projectName, "_raw.RDS"))
```
## Run PCA
```{r}
seurat.object <- RunPCA(seurat.object, ndims.print = 1:5, nfeatures.print = 5, features = all.genes)
```


```{r}
DimPlot(seurat.object, reduction = "pca", group.by = "orig.ident")
```

```{r}
VizDimLoadings(seurat.object, dims = 1:6, nfeatures = 10, reduction = "pca", ncol = 3)

```

Calculate dimensionality
```{r, figures-side, fig.show='hold', out.width="50%"}
ElbowPlot(seurat.object, ndims = 50)
percent.variance(seurat.object@reductions$pca@stdev)
```




## Calculate percent variance
```{r}
tot.var <- percent.variance(seurat.object@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance:", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance:", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance:", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance:", length(which(cumsum(tot.var) <= 95)))

```

# scRNASeq integration	
 neither sctransform nor normalizedata/scaledata gave what looked like reasonable results.

https://satijalab.org/seurat/articles/integration_introduction.html


```{r}
tech.list <- SplitObject(seurat.object, split.by = "orig.ident")
tech.list <- lapply(X = tech.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 10)
})
```


## identify anchors
```{r}

tech.anchors <- FindIntegrationAnchors(object.list = tech.list, anchor.features = all.genes, k.anchor = 20, k.filter = NA)
# tech.integrated <- IntegrateData(anchorset = tech.anchors, dims = 1:30, k.weight = 20, normalization.method = "SCT")
tech.integrated <- IntegrateData(anchorset = tech.anchors, k.weight = 5)
```
## tech.integrated PCA

```{r}
DefaultAssay(tech.integrated) <- "integrated"
tech.integrated <- ScaleData(tech.integrated)
tech.integrated <- RunPCA(tech.integrated, npcs = 30)
DimPlot(tech.integrated, reduction = "pca", group.by = "orig.ident")
```

so it looks like anchor-based integration is much nicer! goign to stick with this approach.

Problem now is, how do we find whether Biomark overlays on 10X



## tech.integrated umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance:", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance:", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance:", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance:", length(which(cumsum(tot.var) <= 95)))
ndims <- length(which(cumsum(tot.var) <= 90))
```

```{r}
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims, reduction = "pca")
tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims, reduction = "pca")
for(x in c(0.5, 1, 1.5, 2, 2.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = ("integrated_snn_res"), x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											cols = color.palette
											) + 
			ggtitle(paste0(projectName, " dim", ndims, "res", gsub("integrated_snn_res", "", meta.col) ))
		plot(myplot)
	}
}
```

Switch ident
```{r}
Idents(tech.integrated) <- "integrated_snn_res.0.5"
```

```{r fig.width = 15, fig.height=20}
DimPlot(tech.integrated, group.by = "clust.ID", pt.size = 2, split.by = "ident", ncol = 4, cols = color.palette, shape.by = "orig.ident") + ggmin::theme_powerpoint()
```


```{r fig.width=20}
DimPlot(tech.integrated, group.by = "orig.ident", pt.size = 2, ncol = 2)
```

```{r fig.width=10, fig.height=20}
DimPlot(tech.integrated, group.by = "orig.ident", pt.size = 2, ncol = 1, split.by = "clust.ID", cols = c("red", "black")) + ggmin::theme_powerpoint()
```




```{r}
png("tech.integrated_biomark167UMAP.res.0.5.png", width = 1600, height = 1600)
DimPlot(tech.integrated, split.by = "ident", group.by = "orig.ident", pt.size = 2, ncol = 5, ) + ggmin::theme_powerpoint()
dev.off()
```


```{r}
png(filename = "tech.integrated_biomark167FeatureFluor.png", height = 1600, width =1600)
FeaturePlot(tech.integrated, features = c("Cd34pos", "Cd135pos", "Cd9pos", "Cd36pos", "Cd41pos", "Cd48pos", "Cd123pos", "Cd150pos", "Cd135pos"), reduction = "pca", cols = c("gray92", "blue"), keep.scale = "all", ncol = 3, pt.size = 2)
dev.off()
```

```{r}
for(i in c("Cd34pos", "Cd135pos", "Cd9pos", "Cd36pos", "Cd41pos", "Cd48pos", "Cd123pos", "Cd150pos", "Cd135pos")){
	my_title <- paste0("tech.integrated_biomark167Fluor_", i, ".png")
	print(my_title)
png(filename = my_title, height = 1000, width = 6200)
my_plot <- FeaturePlot(tech.integrated, features = i, split.by = "ident", reduction = "pca", cols = c("gray92", "blue"), keep.scale = "all", ncol = 5, pt.size = 3)
plot(my_plot)
dev.off()
}
```


```{r}
# for(i in unique(tech.integrated@meta.data$clust.ID)){
	# my_title <- paste0("tech.integrated_biomark167clustID_", i, ".png")
	print(my_title)
png(filename = "tech.integrated_biomark167clustID.png", height = 1000, width = 6200)
my_plot <- DimPlot(tech.integrated, pt.size = 6, group.by = "clust.ID", split.by = "ident", shape.by = "orig.ident") + ggmin::theme_powerpoint()
plot(my_plot)
dev.off()
# }
```


# Biomark Only
## Load biomark data

```{r}
biomark.data <- read.table("biomark167_expressionMatrix.txt", sep = "\t", header = TRUE, row.names = 1)
biomark.data[is.na(biomark.data)] <- 0
biomark.data.T <- data.table::transpose(biomark.data)
colnames(biomark.data.T) <- rownames(biomark.data)
rownames(biomark.data.T) <- colnames(biomark.data)
# fix non-MGI gene names
rownames(biomark.data.T)[rownames(biomark.data.T) == 'Cd117'] <- 'Kit'
rownames(biomark.data.T)[rownames(biomark.data.T) == 'Cd123'] <- 'Il3ra'
rownames(biomark.data.T)[rownames(biomark.data.T) == 'H2.Aa'] <- 'H2-Aa'
rownames(biomark.data.T)[rownames(biomark.data.T) == 'H2.Ab1'] <- 'H2-Ab1'
```

Pull in original cellID info 

```{r}
biomark.order <- read.table("origBiomark_cellIDorder.txt", sep = "\t", header = FALSE, row.names = 1)
```


```{r}
biomark.order["clust.ID"] <- NA
biomark.order[1:49, "clust.ID"] <- 0
biomark.order[50:87, "clust.ID"] <- 17
biomark.order[88:102, "clust.ID"] <- 10
biomark.order[103:117, "clust.ID"] <- 0
biomark.order[118:127, "clust.ID"] <- 3
biomark.order[128:150, "clust.ID"] <- 0
biomark.order[151:189, "clust.ID"] <- 11
biomark.order[190:281, "clust.ID"] <- 0
```

```{r warning=FALSE, message=FALSE}
biomark.meta <- data.frame(stringr::str_split_fixed(rownames(biomark.data), "_", 3)[,2:3], row.names = rownames(biomark.data))
colnames(biomark.meta) <- c("wid", "pid")


biomark.meta['nFeature_RNA'] <- NA
for(cell in colnames(biomark.data.T)){
	ngenes <- length(which(!(biomark.data.T[colnames(biomark.data.T > 0) == cell])))
	biomark.meta[rownames(biomark.meta) == cell, 'nFeature_RNA'] <- ngenes
}
i1 <- match(row.names(biomark.meta), row.names(biomark.order))
biomark.meta[names(biomark.order)[1]] <- lapply(biomark.order[1], `[`, i1)

```

```{r}
# N.B., can't find alternate name for Kng2, so remove
biomark.data.T <- biomark.data.T[!(rownames(biomark.data.T) == "Kng2"), ]
biomark.data.T <- as.sparse(biomark.data.T)
biomark <- CreateSeuratObject(biomark.data.T, project = "Biomark", assay = "RNA", meta.data = biomark.meta)
```





```{r}
biomark <- NormalizeData(biomark, normalization.method = "LogNormalize", scale.factor = 10000)
biomark <- FindVariableFeatures(biomark, selection.method = "vst", nfeatures = 10)
all.genes <- rownames(biomark)
biomark <- ScaleData(biomark, features = all.genes, vars.to.regress = c("nFeature_RNA", "pid", "nCounts_RNA"))
```

### Save merged object
```{r}
# saveRDS(biomark, file = paste0(projectName, "_raw.RDS"))
```
## Run PCA
```{r}
biomark <- RunPCA(biomark, ndims.print = 1:5, nfeatures.print = 5, features = all.genes)
```


```{r}
DimPlot(biomark, reduction = "pca", group.by = "orig.ident")
VizDimLoadings(biomark, dims = 1:6, nfeatures = 10, reduction = "pca", ncol = 3)

```

Calculate dimensionality
```{r}
ElbowPlot(biomark, ndims = 50)
percent.variance(biomark@reductions$pca@stdev)
```




## Calculate percent variance
```{r}
tot.var <- percent.variance(biomark@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance:", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance:", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance:", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance:", length(which(cumsum(tot.var) <= 95)))

```

## Add cluster IDs from Seurat v1

Exported cell IDs for clusters 3, 17, 10, 11 from Seurat v1. Will add these IDs as a metadata column.  
Create column "clust.ID" and populate with 0's. Then import IDs for clusters



## Lets see which cells correspond to our old clusters

```{r}
DimPlot(biomark, reduction = "pca", group.by = "orig.ident") + facet_wrap(~orig.ident)
```


```{r fig.width=15, fig.height=8}
biomark@meta.data$clust.ID <- as.factor(biomark@meta.data$clust.ID)
DimPlot(biomark, group.by = "clust.ID", pt.size = 2, shape.by = "clust.ID")
```



```{r fig.width=15, fig.height = 5}
DimPlot(biomark, split.by = "clust.ID", pt.size = 2, shape.by = "orig.ident")
```






# Setup
## Load libraries required for Seuratv4
```{r message=FALSE, warning=FALSE}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/")
knitr::opts_chunk$set(fig.width = 6, fig.height = 4)
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(clustree)
```
## Source functions
```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/Mouse2Human_idconversion.R")
```

## Set global variables
```{r}
pop.id <- "CMP"
folder.appendix <- "m2"
fav.var <- 90
object.res <- ".0.5"
ident.list <- c("RNA_snn_res.0.5", "RNA_snn_res.1")
folder.id <- paste0(pop.id, folder.appendix)
projectName <- paste0(pop.id, "Subpop_vBiomark")
```

## Store session info
```{r}
sessionInfo.filename <- paste0(projectName, "_sessionInfo.txt")
sink(sessionInfo.filename)
sessionInfo()
sink()
```

*STOP* - testing showed that cells cluster better in seaborn when you include all probe data points. It was an intersting idea to cluster based only on the genes in the biomark probe set, but this doesn't seem to hold up in practice

```{r warning=FALSE, message=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = c(folder.id))
data.object<-Read10X(data_file.list)
seurat.object<- CreateSeuratObject(counts = data.object, min.cells = 3, min.genes = 200, project = projectName)
remove(data.object)
probe.list <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/BiomarkRNAs.txt", header = FALSE)
probe.list <- unlist(as.list(probe.list), use.names = FALSE)
probe.list <- c(probe.list, "Kit", "Il3ra")
# Subset to biomark features
biomark.object <- subset(seurat.object, features = probe.list)
biomark.object[["percent.mt"]] <- PercentageFeatureSet(biomark.object, pattern = "^mt-")
print(paste("original object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- subset(biomark.object, 
												subset = nFeature_RNA >=25 & 
													percent.mt <= 5
)
print(paste("new object:", nrow(biomark.object@meta.data), "cells", sep = " "))
biomark.object <- NormalizeData(biomark.object, normalization.method = "LogNormalize", scale.factor = 10000)
object.downsample <- subset(biomark.object, cells = sample(Cells(biomark.object), 300))
```



## overlay data from clustered seurat.object onto biomark.object
```{r}
seurat.object <- readRDS("~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/CMPSubpop_pctVar90.RDS")
xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "counts"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Counts-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_Data-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

xprsn.mtx <- t(as.data.frame(as.matrix(GetAssayData(object.downsample, slot = "scale.data"))))
object.metadata <- seurat.object@meta.data[rownames(seurat.object@meta.data) %in% rownames(xprsn.mtx), ]
# biomark.export <- xprsn.mtx
biomark.export <- merge(xprsn.mtx, object.metadata[paste0("RNA_snn_res", object.res)], by = 0)
colnames(biomark.export)[colnames(biomark.export) == "Kit"] <- "Cd117"
colnames(biomark.export)[colnames(biomark.export) == "Il3ra"] <- "Cd123"
write.table(biomark.export, file = paste0(projectName, "_ScaleData-BiomarkProbes.txt"), sep = "\t", quote = FALSE, col.names = TRUE, row.names = FALSE)

```




```{r}
biomark.object <- readRDS("~/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/CMPSubpop_pctVar90.RDS")
```








