---
title: "MergeBiomarkWith10X"
output:
  html_document:
    df_print: paged
---

# Updates
## 20211014
Newest idea (C/O Dave) is to take the 10X cells from tech.integrated (the merged Biomark/SingeCell object) and find them on the sc.object. Then add the metadata from  `tech.integrated@metadata$integrated_snn_res`. Running `FindMarkers()` on this parameter might help define population types *without* being biased by the biomark probes.  

## 20211011
Have successfully co-clustered LSK, CMP, and MEP scCells with respective biomark populations. Now have to determine if the fluorophore data supports this.  
Will load the fluorophore data in as metadata, and try to plot the fluor metadata based on the sc/biomark clusters. not sure if the NAs in sc will be a problem, but can subset to the biomark object and try that way.

## 20211007
Gonna crank through LSK and MEP, then go back to the fluorophore data

## 20211005
Compared to archive 1, this version imports Biomark Ct spreadsheets that do not include fluorophore data 

# Setup
```{r setup, message=FALSE, warning=FALSE}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/MergeBiomarkWith10X/")
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(clustree)
# library(scatterplot3d)
library(rgl)
```

## Set global variables
```{r}
pop.of.interest <- "CMP"
cellranger.folder <- "m2"
integrated.var <- 95
integrated.res <- ".1"
integrated.ident.list <- c("integrated_snn_res.0.5", "integrated_snn_res.1", "integrated_snn_res.1.5")
projectName <- paste0("MergeBiomarkWith10X_", pop.of.interest)
sc.pop.var <- 90
sc.pop.res <- ".0.5"
sc.ident.list <- c("RNA_snn_res.0.5", "RNA_snn_res.1")
```
source code
```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/metadata_overlap_stats.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/ColorPalette.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/Mouse2Human_idconversion.R")
```
store session info
```{r }
sink(paste0(projectName, "_sessionInfo.txt"))
sessionInfo()
sink()
```
# Create seurat objects

## Load biomark data
Try to read in Biomark Data

```{r}
biomark.data <- read.table("/Users/heustonef/Desktop/10XGenomicsData/BioMark/Biomark_CT.txt", sep = "\t", header = TRUE, row.names = 1)
biomark.data[is.na(biomark.data)] <- 0

# Limit to biomark pop subset
biomark.data <- biomark.data[grepl(pop.of.interest, rownames(biomark.data)) == TRUE,]
dim(biomark.data)
```
```{r}
# fix non-MGI gene names
colnames(biomark.data)[colnames(biomark.data) == 'Cd117'] <- 'Kit'
colnames(biomark.data)[colnames(biomark.data) == 'Cd123'] <- 'Il3ra'
colnames(biomark.data)[colnames(biomark.data) == 'H2.Aa'] <- 'H2-Aa'
colnames(biomark.data)[colnames(biomark.data) == 'H2.Ab1'] <- 'H2-Ab1'
if("Kng2" %in% biomark.data){biomark.data <- subset(biomark.data, select = -Kng2)}
```
Convert to pseudoexpression data
```{r}
# generate list of reference Ct averages, calculating average for ref genes where ref_gene Ct > 0
ref_list <- c("Actb", "B2m", "Kit")
biomark.data["refCt"] <- NA

for(i in 1:nrow(biomark.data)){
	ref_cts <- c()
	for(j in ref_list){
		if(biomark.data[i, j] > 0){
			ref_cts <- c(ref_cts, biomark.data[i, j])
		}
	}
	biomark.data[i, "refCt"] <- mean(ref_cts)
}

# generate dCt col
dbiomark.data <- biomark.data[, 1:ncol(biomark.data) -1]
dbiomark.data[dbiomark.data == 0] <- NA
dbiomark.data <- dbiomark.data - biomark.data[, "refCt"]

# create pseudo expression matrix
pseudo.biomark <- 2^-dbiomark.data

```





```{r warning=FALSE, message=FALSE}
biomark.meta <- data.frame(stringr::str_split_fixed(rownames(pseudo.biomark), "_", 3)[,2:3], row.names = rownames(pseudo.biomark))
colnames(biomark.meta) <- c("wid", "pid")

pseudo.biomark.T <- data.table::transpose(pseudo.biomark)
colnames(pseudo.biomark.T) <- rownames(pseudo.biomark)
rownames(pseudo.biomark.T) <- colnames(pseudo.biomark)

biomark.meta['nFeature_RNA'] <- NA
for(cell in colnames(pseudo.biomark.T)){
	ngenes <- length(which(!(pseudo.biomark.T[colnames(pseudo.biomark.T > 0) == cell])))
	biomark.meta[rownames(biomark.meta) == cell, 'nFeature_RNA'] <- ngenes
}

```

```{r}
pseudo.biomark.T[is.na(pseudo.biomark.T)] <- 0
pseudo.biomark.T <- as.sparse(pseudo.biomark.T)
biomark.object <- CreateSeuratObject(pseudo.biomark.T, project = "biomark", assay = "RNA", meta.data = biomark.meta)
```
## Thought experiment
```{r}
dbiomark.T <- data.table::transpose(dbiomark.data)
dbiomark.object <- CreateSeuratObject(dbiomark.T, project = "biomark", assay = "RNA", meta.data = biomark.meta)
```



## Load SC data

```{r warning=FALSE}
setwd("~/Desktop/10XGenomicsData/cellRanger/") # temporarily changing wd only works if you run the entire chunk at once
data_file.list <- read_10XGenomics_data(sample.list = paste0(pop.of.interest, cellranger.folder))
object.data <-Read10X(data_file.list)
sc.object<- CreateSeuratObject(counts = object.data, min.cells = 3, min.genes = 200, project = "10X")
```


Keep only genes in both objects

```{r}
shared.features <- intersect(rownames(biomark.object@assays$RNA@data), rownames(sc.object@assays$RNA@data))
sc.subset <- subset(sc.object, features = shared.features)
biomark.object <- subset(biomark.object, features = shared.features)
sc.subset@meta.data["wid"] <- '10X'
sc.subset@meta.data["pid"] <- "p14"

# Check that there's a clean overlap
setdiff(rownames(biomark.object@assays$RNA@data), rownames(sc.subset@assays$RNA@data))
```

Clean up to free memory

```{r}
remove(object.data, biomark.data, biomark.meta)
```

## merge objects
```{r}
merged.object <- merge(sc.subset, biomark.object, project = 'mergebiomark')
merged.object@meta.data$orig.ident[merged.object@meta.data$orig.ident == paste0(pop.of.interest, cellranger.folder)] <- '10X'
merged.object@meta.data$orig.ident[merged.object@meta.data$orig.ident == pop.of.interest] <- 'biomark'
# merged.object@meta.data[is.na(merged.object@meta.data)] <- 0
# merged.object@meta.data$orig.ident <- factor(merged.object@meta.data$orig.ident)
# merged.object@meta.data$pid <- factor(merged.object@meta.data$pid)
# merged.object@meta.data$wid <- factor(merged.object@meta.data$wid)
merged.object
```

# Traditional normalize and scaling approaches
trying sctransform and normalizeData/scaledata
```{r}
# merged.object <- SCTransform(merged.object, vars.to.regress = c("wid"))
merged.object <- NormalizeData(merged.object, normalization.method = "LogNormalize", scale.factor = 10000)
merged.object <- FindVariableFeatures(merged.object, selection.method = "vst", nfeatures = 10)
```

Scale data (linear transformation)

```{r}
all.genes <- rownames(merged.object)
merged.object <- ScaleData(merged.object, features = all.genes, vars.to.regress = c("wid", "pid", "orig.ident"))
```

### Save merged object
```{r}
# saveRDS(merged.object, file = paste0(projectName, "_raw.RDS"))
```
## Run PCA
```{r}
merged.object <- RunPCA(merged.object, ndims.print = 1:5, nfeatures.print = 5, features = all.genes)
```


```{r}
DimPlot(merged.object, reduction = "pca", group.by = "orig.ident", cols = c("black", "red"), pt.size = 1)
png(filename = paste0(projectName, "merged_PCA.png"), height = 800, width = 800)
DimPlot(merged.object, reduction = "pca", group.by = "orig.ident", cols = c("black", "red"), pt.size = 2)
dev.off()
```

```{r}
VizDimLoadings(merged.object, dims = 1:6, nfeatures = 10, reduction = "pca", ncol = 3)

```

Calculate dimensionality
```{r, figures-side, fig.show='hold', out.width="50%"}
ElbowPlot(merged.object, ndims = 50)
percent.variance(merged.object@reductions$pca@stdev)
```




## Calculate percent variance
```{r}
tot.var <- percent.variance(merged.object@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance:", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance:", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance:", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance:", length(which(cumsum(tot.var) <= 95)))

```

# scRNASeq integration	
neither sctransform nor normalizedata/scaledata gave what looked like reasonable results.

https://satijalab.org/seurat/articles/integration_introduction.html


```{r}
tech.list <- c(sc.subset, biomark.object)
tech.list <- lapply(X = tech.list, FUN = function(x) {
	x <- NormalizeData(x)
	x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 10)
})
```


## identify anchors
```{r}
tech.anchors <- FindIntegrationAnchors(object.list = tech.list, anchor.features = all.genes, k.anchor = 20, k.filter = NA)
# tech.integrated <- IntegrateData(anchorset = tech.anchors, dims = 1:30, k.weight = 20, normalization.method = "SCT")
tech.integrated <- IntegrateData(anchorset = tech.anchors, k.weight = 5)
```

## tech.integrated PCA

```{r}
DefaultAssay(tech.integrated) <- "integrated"
tech.integrated <- ScaleData(tech.integrated)
tech.integrated <- RunPCA(tech.integrated, npcs = 30)
DimPlot(tech.integrated, reduction = "pca", group.by = "orig.ident", pt.size = 2, cols = c("red", "black"))

png(filename = paste0(projectName, "anchored_PCA.png"), height = 800, width = 800, bg = "transparent")
DimPlot(tech.integrated, reduction = "pca", group.by = "orig.ident", pt.size = 2, cols = c("red", "black"), combine = FALSE)
dev.off()
```

so it looks like anchor-based integration is much nicer! goign to stick with this approach.



Number of PCs describing X% of variance

```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
paste0("Num pcs for 80% variance: ", length(which(cumsum(tot.var) <= 80)))
paste0("Num pcs for 85% variance: ", length(which(cumsum(tot.var) <= 85)))
paste0("Num pcs for 90% variance: ", length(which(cumsum(tot.var) <= 90)))
paste0("Num pcs for 95% variance: ", length(which(cumsum(tot.var) <= 95)))

```


# ID clusters based on different variances
## Describe 80% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 80
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
find clusters

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```







## Describe 85% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 85
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```






## Describe 90% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 90
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r}
for(x in c(0.5, 1, 1.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```

```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```

### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```






## Describe 95% of variance 

### Neighborhood and umap
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
pct.var <- 95
ndims <- length(which(cumsum(tot.var) <= pct.var))
print(ndims)

tech.integrated <- FindNeighbors(tech.integrated, dims = 1:ndims)
tech.integrated <- FindClusters(tech.integrated, resolution = 0.5)
tech.integrated <- RunUMAP(tech.integrated, dims = 1: ndims)
```
Plot UMAP

```{r warning=FALSE, results='hide'}
for(x in c(0.5, 1, 1.5, 2, 2.5)){
	tech.integrated <- FindClusters(tech.integrated, resolution = x)
}
```

```{r}
for (meta.col in colnames(tech.integrated@meta.data)){
	if(grepl(pattern = "integrated_snn_res", x = meta.col)==TRUE | grepl(pattern = "orig.ident", x = meta.col)==TRUE){
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col) ))
		plot(myplot)
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		myplot <- DimPlot(tech.integrated, 
											group.by = meta.col,
											reduction = "umap", 
											pt.size = 1,
											cols = color.palette) + 
			facet_wrap(meta.col) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col)))
		
		png(filename = paste0(projectName, "_pctVar", pct.var, "res.", gsub("integrated_snn_res.", "", meta.col), "-umap_FacetRes.png"), height = 800, width = 800)
		plot(myplot)
		dev.off()
		
	}
}
```
### Evaluate cluster stability
```{r fig.height=5}
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") + 
	scale_color_continuous(low = 'red3', high = 'white')
png(filename = paste0(projectName, "_pctVar", pct.var, "-clustree.png"), height = 800, width = 1600)
clustree(tech.integrated, prefix = "integrated_snn_res.", node_colour = "sc3_stability") +
	scale_color_continuous(low = 'red3', high = 'white')
dev.off()
```
```{r}
saveRDS(tech.integrated, file = paste0(projectName, "_pctVar", pct.var, ".RDS"))
```


# Pick set of conditions to investigate further
Now that have checked multiple pct.vars and integrated.res's, pick the favourite combination and do more detailed characerization of the. This includes:
* Stats on each cluster
* Find markers
* Export lists for GSEA


## Reload object 
```{r}
tech.integrated <- readRDS(paste0(projectName, "_pctVar", integrated.var, ".RDS"))
tech.integrated <- SetIdent(tech.integrated, value = paste0("integrated_snn_res", integrated.res))
```
for each resolution, number/percentage of cells in each cluster?
```{r}
res.list <- c("integrated_snn_res.0.5", "integrated_snn_res.1")
for(current_res in res.list){
	tech.integrated@meta.data$orig.ident[tech.integrated@meta.data$orig.ident == paste0(pop.of.interest, cellranger.folder)] <- "10X"
	tech.integrated@meta.data$orig.ident[tech.integrated@meta.data$orig.ident == paste0(pop.of.interest)] <- "biomark"
	tech.integrated@meta.data$orig.ident <- factor(tech.integrated@meta.data$orig.ident)
	
	
	cluster_ids <- sort(unique(tech.integrated@meta.data[,current_res]))
	counts_df <- data.frame(matrix(nrow = length(cluster_ids), ncol = 4))
	rownames(counts_df) <- cluster_ids
	colnames(counts_df) <- c("10X", "biomark", "pct10X", "pctbiomark")
	tot.10X <- nrow(tech.integrated@meta.data[tech.integrated@meta.data$orig.ident == "10X", ])
	tot.biomark <- nrow(tech.integrated@meta.data[tech.integrated@meta.data$orig.ident == "biomark", ])
	for(id in cluster_ids){
		cell_value <- nrow(tech.integrated@meta.data[(tech.integrated@meta.data[current_res] == id) & 
																								 	(tech.integrated@meta.data$orig.ident == "10X"),])
		counts_df[id, "10X"] = cell_value
		counts_df[id, "pct10X"] = round(cell_value/tot.10X*100, 2)
		cell_value <- nrow(tech.integrated@meta.data[(tech.integrated@meta.data[current_res] == id) & 
																								 	(tech.integrated@meta.data$orig.ident == "biomark"),])
		counts_df[id, "biomark"] = cell_value
		counts_df[id, "pctbiomark"] = round(cell_value/tot.biomark*100, 2)
	}
	try(xlsx::write.xlsx(x = counts_df, 
											 file = paste0(projectName, "_pctVar", pct.var, "_stats.xlsx"), 
											 sheetName = current_res, 
											 col.names = TRUE, 
											 row.names = TRUE, 
											 append = TRUE)
	)
	counts_df
}
```






## Create `FindAllMarkers()` lists for GSEA
```{r}
integrated.ident.list <- c(paste0("integrated_snn_res", integrated.res))

for(tested.ident in integrated.ident.list){
	Idents(tech.integrated) <- tested.ident
	all.markers <- FindAllMarkers(tech.integrated)
	try(xlsx::write.xlsx(x = all.markers[,c("avg_log2FC", "p_val_adj", "cluster", "gene")], 
									 file = paste0(projectName, "_pctVar", pct.var, "_allres.xlsx"), 
									 sheetName = tested.ident, 
									 col.names = TRUE, 
									 row.names = FALSE, 
									 append = TRUE)
			)
}
```

## Map HGNC symbols
```{r}
tech.integrated <- SetIdent(tech.integrated, value = paste0("integrated_snn_res", integrated.res))
integrated.res.allmarkers <- FindAllMarkers(tech.integrated)
Mouse2HumanTable <- Mouse2Human(integrated.res.allmarkers$gene)

HGNC <- with(Mouse2HumanTable, Mouse2HumanTable$HGNC[match(integrated.res.allmarkers$gene, Mouse2HumanTable$MGI)])
head(integrated.res.allmarkers)
integrated.res.allmarkers$HGNC <- HGNC
# tail(integrated.res.allmarkers)
sig.res <- integrated.res.allmarkers[integrated.res.allmarkers$p_val_adj <= 0.05, ]
sig.res <- sig.res[c("avg_log2FC", "HGNC", "cluster")]
sig.res <- sig.res[!(sig.res$HGNC == "" | is.na(sig.res$HGNC)),] # GSEA will fail if there are any blanks or NAs in the table
sig.res <- sig.res[]

for(cluster in unique(sig.res$cluster)){
	print(paste("writing cluster", cluster))
	new.table <- sig.res[sig.res$cluster == cluster, c("HGNC", "avg_log2FC")]
	new.table <- new.table[order(-new.table$avg_log2FC), ]
	try(dir.create(paste0("RankList_findAll_hgnc/"), showWarnings = FALSE))
	try(write.table(new.table, file = paste0("RankList_findAll_hgnc/", projectName, "pctVar", pct.var, "res", integrated.res, "cluster", cluster, ".rnk"), quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t", ))
	
}
```

# Merge index sorting data

## load object
```{r}
tech.integrated <- readRDS(paste0(projectName, "_pctVar", integrated.var, ".RDS"))
tech.integrated <- SetIdent(tech.integrated, value = paste0("integrated_snn_res", integrated.res))
tech.integrated@meta.data$orig.ident[tech.integrated@meta.data$orig.ident == paste0(pop.of.interest, cellranger.folder)] <- "10X"
tech.integrated@meta.data$orig.ident[tech.integrated@meta.data$orig.ident == paste0(pop.of.interest)] <- "biomark"
tech.integrated@meta.data$orig.ident <- factor(tech.integrated@meta.data$orig.ident)
```

## load index sort data
```{r}
mfi <- read.csv(file = paste0(pop.of.interest, "_mfi_unNormalized.txt"), sep = "\t", row.names = 1, skipNul = TRUE)
head(mfi)
```

sort_dict = {'010319':['p1', 'p3'], '121218': ['p2', 'p4'], '020119': ['p6', 'p7'], '090618': ['p12', 'p13']}
normalize based on sort data  
Plot to check distribution
```{r}
for(colname in colnames(mfi[,grepl("pos$", names(mfi))])){
	h <- hist(mfi[, colname], plot = FALSE) 
	h$density = h$counts/sum(h$counts)*100
	plot(h,freq=FALSE, main = colname, xlab = "mfi", ylab = "% Cells")
	}
```
So some distributions are gaussian, and some are poisson. Best bet is probaby just to normalize everyone.  
> Min-max normalization: guarantees the same scale, but is thrown by outliers  
> Z score normalization: does not guarantee the same scale, but handles outliers better  

Flow is full of outliers. Going to z-score normalize

```{r}
fluor.cols <- names(mfi[, grepl("pos$", names(mfi))])
mfi.norm <- data.frame(matrix(nrow = 0, ncol = length(fluor.cols)))
colnames(mfi.norm) <- fluor.cols
for(sortdate in unique(mfi$sortdate)){
	mfi.subset <- scale(mfi[mfi$sortdate == sortdate, fluor.cols])
	mfi.norm <- rbind(mfi.norm, mfi.subset)
}
mfi.norm["pid"] <- do.call(rbind, strsplit(rownames(mfi.norm), '_'))[,3]
```
Add metadata column
```{r}
tech.integrated <- AddMetaData(tech.integrated, metadata = mfi.norm)
```
now plot some stuff
```{r}
biomark.integrated <- subset(tech.integrated, subset = orig.ident == "biomark")
```
```{r fig.height=12, fig.width=5}
RidgePlot(biomark.integrated, features = fluor.cols, cols = color.palette, ncol = 1)
```
```{r}
png(filename = paste0(projectName, "_pctVar", integrated.var, "res", integrated.res, "-fluor.png"), height = 2400, width = 800)
RidgePlot(biomark.integrated, features = fluor.cols, cols = color.palette, ncol = 1)
dev.off()
```
plot umaps
```{r}
for(integrated.col in integrated.ident.list){
		plot.title <- paste0(projectName, "_pctVar", pct.var, "-mergedEnrichment_OrigIdent-Facet-integratedRes.", gsub("integrated_snn_res.", "", integrated.col))
		myplot <- DimPlot(tech.integrated, 
											group.by = "orig.ident",
											split.by = integrated.col, 
											reduction = "umap", 
											pt.size = 1.5,
											ncol = round(sqrt(length(levels(tech.integrated@meta.data[[sc.col]])))),
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "-", integrated.col, " faceted by ", sc.col))
		plot(myplot)
		
		png(filename = paste0(projectName, "_pctVar", pct.var,
													"-mergedEnrichment_OrigIdent-Facet-integratedRes.", gsub("integrated_snn_res.", "", integrated.col),
													".png"),
				height = 1600, width = 1600)
		plot(myplot)
		dev.off()

}
```

# Add tech.integrated cluster IDs to sc.object
## Load analyzed sc.object from Indivdual_SubpopAnalysis
Use previously determined optimal percent variance to describe (`sc.pop.var`) and resolution (`sc.pop.res`) from notebook "Individual_SubpopAnalysis" and clustree. Will increase resolution if can get improved coverage of integrated_snn clusters
```{r}
seurat.object <- readRDS(paste0("/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/", pop.of.interest, "Subpop_pctVar", sc.pop.var, ".RDS"))
seurat.object
```
Get cluster IDs from merged SC and Biomark seurat object (tech.integrated) and add them back to the original sc.object.
```{r}
seurat.object <- AddMetaData(seurat.object, metadata = tech.integrated@meta.data[, integrated.ident.list])
```
Find the percent of integrated_snn_res.X cells for each RNA_snn_res.in the seurat.object
Test: what percentage of each new clusterID matches one of the older clusters?

```{r}
test.col.pattern <- "integrated_snn_res"
target.col.pattern <- "RNA_snn_res"
xlsx_filename <- paste0(projectName, "_pctVar", pct.var, "_mergedEnrichmentStats.xlsx")

metadata_overlap_stats(seurat.object = seurat.object, 
											 target.col.pattern = target.col.pattern, 
											 test.col.pattern = test.col.pattern, 
											 print_to_screen = TRUE,
											 xlsx_filename = xlsx_filename)
```

plot overlays
```{r}
for(integrated.col in integrated.ident.list){
	for(sc.col in sc.ident.list){
		plot.title <- paste0(projectName, "_pctVar", pct.var, "-", integrated.col, " faceted by ", sc.col)
		myplot <- DimPlot(seurat.object, 
											group.by = integrated.col,
											split.by = sc.col, 
											reduction = "umap", 
											pt.size = 1.5,
											ncol = round(sqrt(length(levels(seurat.object@meta.data[[sc.col]])))),
											cols = color.palette) + 
			ggtitle(paste0(projectName, "_pctVar", pct.var, "-", integrated.col, " faceted by ", sc.col))
		
		png(filename = paste0(projectName, "_pctVar", pct.var, 
													"-integratedRes.", gsub("integrated_snn_res.", "", integrated.col), 
													"mergedEnrichment-Facet_scRes.", gsub("RNA_snn_res.", "", sc.col), 
													".png"), 
				height = 1600, width = 1600)
		plot(myplot)
		dev.off()

	}
}
```

Save object	
```{r}
# saveRDS(seurat.object, file = paste0("/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/IndividualPops/", pop.of.interest, "Subpop_pctVar", sc.pop.var, ".RDS"))
```
# DGE
```{r}
for(tested.ident in integrated.ident.list){
	Idents(seurat.object) <- tested.ident
	all.markers <- FindAllMarkers(seurat.object)
	xlsx::write.xlsx(x = all.markers[,c("avg_log2FC", "p_val_adj", "cluster", "gene")], 
									 file = paste0(projectName, "_FindALLMarkers_pctVar", pct.var, "_allres.xlsx"), 
									 sheetName = tested.ident, 
									 col.names = TRUE, 
									 row.names = FALSE, 
									 append = TRUE)
}
```
















# Plot 3D
Plot in 3d
```{r}
tot.var <- percent.variance(tech.integrated@reductions$pca@stdev, plot.var = FALSE, return.val = TRUE)
ndims <- length(which(cumsum(tot.var) <= integrated.var))
tech.integrated.3d <- tech.integrated
tech.integrated.3d <- FindNeighbors(tech.integrated.3d, dims = 1:ndims)
tech.integrated.3d <- FindClusters(tech.integrated.3d, resolution = as.numeric(integrated.res))
tech.integrated.3d <- RunUMAP(tech.integrated.3d, dims = 1: ndims, n.components = 3)
```
Plot 3D UMAP

```{r}
tech.integrated.3d <- SetIdent(tech.integrated.3d, value = paste0("integrated_snn_res", integrated.res))
my_object <- tech.integrated.3d@reductions$umap@cell.embeddings

# color.palette @ res.0.5 has 15 levels. Change color.palette[11] to a different color so can add the biomark cells as large black dots afterwards
color.palette[11] <- "mediumpurple1"
par3d("windowRect" = c(122, 44, 950, 929))
plot3d(x=my_object[,1], y=my_object[,2], z=my_object[,3], 
			 col = color.palette[tech.integrated.3d@active.ident], 
			 box=FALSE, axes=FALSE, xlab = "", ylab = "", zlab = "")
rgl.lines(c(-10, 10), c(0, 0), c(0, 0), color = "black")
rgl.lines(c(0, 0), c(-10, 10), c(0, 0), color = "red")
rgl.lines(c(0, 0), c(0, 0), c(-10, 10), color = "green")

biomark.embeddings <- subset(tech.integrated.3d, subset = orig.ident == "biomark")
biomark.embeddings <- biomark.embeddings@reductions$umap@cell.embeddings
plot3d(x = biomark.embeddings[,1], y = biomark.embeddings[,2], z = biomark.embeddings[,3],
			 col = "black", size = 10,
			 add = TRUE)
rgl.viewpoint(zoom = .5)
movie3d(spin3d(axis = c(1, 1, 1), rpm = 6), duration = 10, dir = getwd(), clean = TRUE, webshot = TRUE, movie = "MergeBiomarkWith10X_CMP_pctVar95res.1.5")
```

# Find Seurat v1 cells
## Add cluster IDs from Seurat v1
Exported cell IDs for clusters 3, 17, 10, 11 from Seurat v1. Will add these IDs as a metadata column.  
Create column "clust.ID" and populate with 0's. Then import IDs for clusters
```{r}
clust3.cells <- read.table(file = "/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/Seuratv1_clusterCellIDs/cluster3cellIDs.txt", col.names = "clust03")
clust3.cells <- sapply(clust3.cells, function(x) paste0(gsub("CMP", "CMPm2", x), "-1"))
clust17.cells <- read.table(file = "/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/Seuratv1_clusterCellIDs/cluster17cellIDs.txt", col.names = "clust17")
clust17.cells <- sapply(clust17.cells, function(x) paste0(gsub("CMP", "CMPm2", x), "-1"))
clust10.cells <- read.table(file = "/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/Seuratv1_clusterCellIDs/cluster10cellIDs.txt", col.names = "clust10")
clust10.cells <- sapply(clust10.cells, function(x) paste0(gsub("CMP", "CMPm2", x), "-1"))
clust11.cells <- read.table(file = "/Users/heustonef/Desktop/10XGenomicsData/msAggr_scRNASeq/Seuratv1_clusterCellIDs/cluster11cellIDs.txt", col.names = "clust11")
clust11.cells <- sapply(clust11.cells, function(x) paste0(gsub("CMP", "CMPm2", x), "-1"))
```

## map cluster ids from Seuratv1
```{r}
seurat.object@meta.data$clust.ID[rownames(seurat.object@meta.data) %in% clust3.cells] <- 3
seurat.object@meta.data$clust.ID[rownames(seurat.object@meta.data) %in% clust17.cells] <- 17
seurat.object@meta.data$clust.ID[rownames(seurat.object@meta.data) %in% clust10.cells] <- 10
seurat.object@meta.data$clust.ID[rownames(seurat.object@meta.data) %in% clust11.cells] <- 11
seurat.object@meta.data[is.na(seurat.object@meta.data)] <- 0
seurat.object@meta.data$clust.ID <- as.factor(seurat.object@meta.data$clust.ID)
```
## plot something
```{r}
DimPlot(tech.integrated, group.by = "clust.ID", split.by = "integrated_snn_res.1.5", cols = c("black", "green", "blue", "red", "orange"), ncol = 4)
```


```{r}
test.col.pattern <- "RNA_snn_res.1"
target.col.pattern <- "integrated_snn_res"
xlsx_filename <- paste0(projectName, "_pctVar", pct.var, "_mergedEnrichmentStats_Sv4B.xlsx")

metadata_overlap_stats(seurat.object = seurat.object, 
											 target.col.pattern = target.col.pattern, 
											 test.col.pattern = test.col.pattern, 
											 print_to_screen = TRUE,
											 xlsx_filename = xlsx_filename)

```

## map cluster ids from Seuratv1
```{r}
tech.integrated@meta.data$clust.ID[rownames(tech.integrated@meta.data) %in% clust3.cells] <- 3
tech.integrated@meta.data$clust.ID[rownames(tech.integrated@meta.data) %in% clust17.cells] <- 17
tech.integrated@meta.data$clust.ID[rownames(tech.integrated@meta.data) %in% clust10.cells] <- 10
tech.integrated@meta.data$clust.ID[rownames(tech.integrated@meta.data) %in% clust11.cells] <- 11
tech.integrated@meta.data[is.na(tech.integrated@meta.data)] <- 0
tech.integrated@meta.data$clust.ID <- as.factor(tech.integrated@meta.data$clust.ID)
```
## plot something
```{r}
DimPlot(tech.integrated, group.by = "clust.ID", split.by = "integrated_snn_res.1.5", cols = c("black", "green", "blue", "red", "orange"), ncol = 11)
```

```{r}
DimPlot(seurat.object, group.by = "clust.ID", cols = c("black", "green", "blue", "red", "orange"))
DimPlot(seurat.object, group.by = "clust.ID", split.by = "RNA_snn_res.1", cols = c("black", "green", "blue", "red", "orange"), ncol = 5)

png(filename = paste0(projectName, "_pctVar", pct.var, "-Sv1ClustID_on_Sv4.png"), height = 800, width = 800)
DimPlot(seurat.object, group.by = "clust.ID", cols = c("black", "green", "blue", "red", "orange"), pt.size = 1)
dev.off()
png(filename = paste0(projectName, "_pctVar", pct.var, "-Sv1ClustID_on_Sv4-facetSv4.png"), height = 800, width = 800)
DimPlot(seurat.object, group.by = "clust.ID", split.by = "RNA_snn_res.1", cols = c("black", "green", "blue", "red", "orange"), ncol = 5) + NoAxes() + NoLegend()
dev.off()
```

## generate emphasis colorpalette
```{r}
emphasis.palette <- rep("black", times = 25)
```

```{r}
emphasis.palette.Sv4 <- emphasis.palette
emphasis.palette.Sv4[c(4, 6, 7, 8)] <- c("green", "orange", "red", "blue")
DimPlot(seurat.object, group.by = "RNA_snn_res.1", split.by = "integrated_snn_res.1.5", cols = emphasis.palette.Sv4, ncol = 5)

png(filename = paste0(projectName, "_pctVar", pct.var, "-Sv4ClustID_on_Sv4B-facetSv4B.png"), height = 800, width = 800)
DimPlot(seurat.object, group.by = "RNA_snn_res.1", split.by = "integrated_snn_res.1.5", cols = emphasis.palette.Sv4, ncol = 5) + NoAxes()
dev.off()
```











