---
title: "pseudobulk_CMP"
output: html_notebook
---

# Purpose
Determine _ranked spearman correlation_ between single cell RNASeq (scRNASeq) and bulk RNASeq (bRNASeq)

## Method
1. Perform pseudo-bulk analysis from scRNASeq of population (i.e., CMP)
2. Download bRNASeq of same population (has 2 replicates)
3. Ranked spearman correlation between bRNASeq replicates
4. Ranked spearman correlation between each bRNASeq replicate and pseudobulk scRNASeq sample
5. Ranked spearman correlation between combined bRNASeq replicates and pseudobulk scRNASeq sample

## Update
### 2021.09.13
Finally got a method working for CMP vs. TotalSeq CMP (see pseudobulk_CMP.RMD). Will now clean up notebook and comparisons for both TotalSeq and ScriptSeq, and for all LSK, CMP, MEP, and GMP populations.  
Following a new [tutorial][https://jdblischak.github.io/singlecell-qtl/sc-vs-bulk.html]. 



# Notebook setup

```{r setup}
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/msAggr_scRNASeq/pseudobulk_vs_bulk/")
```

## Load libraries
```{r}
library(dplyr)
library(Seurat)
library(ggplot2)
```

## Set global variables
```{r}
projectName <- "pseudobulk_allPops"
comparison.pop <- "CMP"
export.graphs <- TRUE
```

## Store session info
```{r}
sessionInfo.filename <- paste0(projectName, "_sessionInfo.txt")
sink(sessionInfo.filename)
sessionInfo()
sink()
```

## Load local scripts
```{r}
# source("../RFunctions/read_10XGenomics_data.R")
# source("../RFunctions/PercentVariance.R")
# source("../RFunctions/Mouse2Human_idconversion.R")
source ("../RFunctions/ColorPalette.R")
```

# Load scRNASeq object
##Load raw object
Has feature and mt filters applied
```{r}
seurat.object <- readRDS(paste0("~/Desktop/10XGenomicsData/msAggr_scRNASeq/raw_seuratObjects/", comparison.pop, "_raw.rds"))
```

# Calculate cpm 
from single cell umi matrix
```{r}
sc.mtx <- data.frame(rowSums(seurat.object@assays$RNA@counts))
sc.mtx <- sc.mtx/colSums(sc.mtx)*1000000
```
_Tutorial states that only quantity directly cmoprable between bulk and scRNA-Seq is relative abundance (cites Pachter 2001)._

# Load bRNASeq data
## read cmp bulk TPM
```{r}
bulk.subfolders <- c("../../bRNASeq/TotalSeq/", "../../bRNASeq/ScriptSeq/")
bulk.path <- bulk.subfolders[1]
file_list <- list.files(path = bulk.path, pattern = comparison.pop)
bulk.pop <- read.table(file = paste0(bulk.path, file_list[1]), header = TRUE)[1]

for(file in file_list){
	print(file)
	bulk.pop <- merge(bulk.pop, read.table(file = paste0(bulk.path, file), header = TRUE, sep = "\t")[, c(1,5)], by = "gene_id")
}

#calculate average tpm
bulk.pop[comparison.pop] <- rowMeans(x = bulk.pop[, c(2, 3)])

# clean and check df
bulk.pop <- bulk.pop[, -c(2, 3)]
head(bulk.pop)
```
```{r}
#calc log_rho's
bulk.rho <- log((bulk.pop[comparison.pop] + 1), 2) - log(colSums((bulk.pop[comparison.pop])+ 1), 2)
rownames(bulk.rho) <- bulk.pop$gene_id
colnames(bulk.rho) <- "bulk"
sc.rho <- log((sc.mtx + 1), 2) - log(colSums((sc.mtx+1)), 2)
colnames(sc.rho) <- "sc"
```

check that all values are numeric
```{r}
print(is.numeric(bulk.rho[,1]))
print(is.numeric(sc.rho[,1]))
```
## create concordance df
```{r}
# drop NAs
concordance.df <- merge(bulk.rho, sc.rho, by = "row.names", all = TRUE)
concordance.df <- na.omit(concordance.df)
rownames(concordance.df) <- concordance.df$Row.names
concordance.df <- concordance.df[, c(2, 3)]
```
```{r}
spearman.coeff <- cor(concordance.df, method = "spearman")[1, 2]
spearman.coeff
```
## plot data
```{r}
cor.plot <- ggplot(data = concordance.df, aes(x = bulk, y = sc)) + 
	geom_point()  +
	ggtitle(paste0("concordance in ", comparison.pop, " spearman coeff = ", round(spearman.coeff, digits = 3)))
plot(cor.plot)
```

## print images
```{r}
if(export.graphs == TRUE){
	png(filename = paste0(comparison.pop, "_spearmanPlot.png"), height = 480, width = 800)
	plot(cor.plot)
	dev.off()
} else{
	print("Export graphs set to FALSE")
}
```





# Automate through bRNASeq data
```{r}
bulk.folders <- c("TotalSeq", "ScriptSeq")

for(sequencing.folder in bulk.folders){
	bulk.subfolders <- c(paste0("~/Desktop/10XGenomicsData/bRNASeq/", sequencing.folder, "/"))
	
	for(bulk.path in bulk.subfolders){
		bulk.path <- bulk.subfolders[1]
		file_list <- list.files(path = bulk.path, pattern = comparison.pop)
		bulk.pop <- read.table(file = paste0(bulk.path, file_list[1]), header = TRUE)[1]
		
		for(file in file_list){
			print(file)
			bulk.pop <- merge(bulk.pop, read.table(file = paste0(bulk.path, file), header = TRUE, sep = "\t")[, c(1,5)], by = "gene_id")
		}
		
		#calculate average tpm
		bulk.pop[comparison.pop] <- rowMeans(x = bulk.pop[, c(2, 3)])
		
		# clean and check df
		bulk.pop <- bulk.pop[, -c(2, 3)]
		head(bulk.pop)
		
		
		#calc log_rho's
		bulk.rho <- log((bulk.pop[comparison.pop] + 1), 2) - log(colSums((bulk.pop[comparison.pop])+ 1), 2)
		rownames(bulk.rho) <- bulk.pop$gene_id
		colnames(bulk.rho) <- "bulk"
		sc.rho <- log((sc.mtx + 1), 2) - log(colSums((sc.mtx+1)), 2)
		colnames(sc.rho) <- "sc"
		
		# check all entries are numeric
		print(is.numeric(bulk.rho[,1]))
		print(is.numeric(sc.rho[,1]))
		
		
		# drop NAs
		concordance.df <- merge(bulk.rho, sc.rho, by = "row.names", all = TRUE)
		concordance.df <- na.omit(concordance.df)
		rownames(concordance.df) <- concordance.df$Row.names
		concordance.df <- concordance.df[, c(2, 3)]
		
		# spearman plot
		spearman.coeff <- cor(concordance.df, method = "spearman")[1, 2]
		print(spearman.coeff)
		
		
		cor.plot <- ggplot(data = concordance.df, aes(x = bulk, y = sc)) + 
			geom_point()  +
			ggtitle(paste0("concordance in ", comparison.pop, " vs ", sequencing.folder, ": spearman coeff = ", round(spearman.coeff, digits = 3)))
		plot(cor.plot)
		
		
		if(export.graphs == TRUE){
			png(filename = paste0(comparison.pop, "_vs", sequencing.folder,"-spearmanPlot.png"), height = 480, width = 800)
			plot(cor.plot)
			dev.off()
		} else{
			print("Export graphs set to FALSE")
		}
	}
	
}
```



