---
title: "Create Raw Seurat Objects"
output: html_notebook
---

# Purpose
Purpose of this notebook is simply to generate raw seurat objects for analysis in other programs

# Notebook setup

Creating new pipeline using seurat v4.0.2 available 2021.06.08

Load libraries required for Seuratv4

```{r setup	}
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
knitr::opts_knit$set(root.dir = "~/Desktop/10XGenomicsData/cellRanger/")
```

# Set global variables
```{r}
projectName <- "GenerateRawSeuratObject"
```

store session info
```{r }
sessionInfo.filename <- paste0(projectName, "_sessionInfo.txt")
sink(sessionInfo.filename)
sessionInfo()
sink()
```


```{r}
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/read_10XGenomics_data.R")
source("~/Desktop/10XGenomicsData/msAggr_scRNASeq/RFunctions/PercentVariance.R")
```


```{r warning=FALSE}
data_file.list <- read_10XGenomics_data(sample.list = "CMPm2")
object.data <-Read10X(data_file.list)
seurat.object<- CreateSeuratObject(counts = object.data, min.cells = 3, min.genes = 200, project = projectName)

# Clean up to free memory

remove(object.data)

# Add mitochondrial metadata and plot some basic features

seurat.object[["percent.mt"]] <- PercentageFeatureSet(seurat.object, pattern = "^mt-")
VlnPlot(seurat.object, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size = 0, fill.by = 'orig.ident', )

plot1 <- FeatureScatter(seurat.object, feature1 = "nCount_RNA", feature2 = "percent.mt", group.by = "orig.ident", pt.size = 0.01)
plot2 <- FeatureScatter(seurat.object, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", group.by = "orig.ident", pt.size = 0.01)
plot1 + plot2

# remove low quality cells
# require: nFeature_RNA between 200 and 4000 (inclusive)
# require: percent.mt <=5

print(paste("original object:", nrow(seurat.object@meta.data), "cells", sep = " "))
seurat.object <- subset(seurat.object, 
												subset = nFeature_RNA >=200 & 
													nFeature_RNA <= 4000 & 
													percent.mt <= 5
												)
print(paste("new object:", nrow(seurat.object@meta.data), "cells", sep = " "))

# Save raw object

saveRDS(seurat.object, file = paste0(projectName, "_raw.RDS"))
```


